spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/CassandraJoinRDD.scala

```diff
@@ -0,0 +1,152 @@
+                                                val connector: CassandraConnector,
+                                                val columnNames: ColumnSelector = AllColumns,
+                                                val joinColumns: ColumnSelector = PartitionKeyColumns,
+                                                val where: CqlWhereClause = CqlWhereClause.empty,
+                                                val readConf: ReadConf = ReadConf())
+                                               (implicit oldTag: ClassTag[O], val rct: ClassTag[N],
+                                                @transient val rwf: RowWriterFactory[O], @transient val rtf: RowReaderFactory[N])
+  extends CassandraRDD[(O, N)](prev.sparkContext, prev.dependencies) with CassandraReader[N] {
+
+  //Make sure copy operations make new CJRDDs and not CRDDs
+  override protected def copy(columnNames: ColumnSelector = columnNames,
+                              where: CqlWhereClause = where,
+                              readConf: ReadConf = readConf, connector: CassandraConnector = connector) =
+    new CassandraJoinRDD[O, N](prev, keyspaceName, tableName, connector, columnNames, joinColumns, where, readConf).asInstanceOf[this.type]
+
+  lazy val joinColumnNames: Seq[NamedColumnRef] = joinColumns match {
+    case AllColumns => throw new IllegalArgumentException("Unable to join against all columns in a Cassandra Table. Only primary key columns allowed")
+    case PartitionKeyColumns => tableDef.partitionKey.map(col => col.columnName: NamedColumnRef).toSeq
+    case SomeColumns(cs@_*) => {
+      checkColumnsExistence(cs)
+      checkValidJoin(cs)
+    }
+  }
+
+  protected def checkValidJoin(columns: Seq[NamedColumnRef]): Seq[NamedColumnRef] = {
+    val allColumnNames = tableDef.allColumns.map(_.columnName).toSet
+    val regularColumnNames = tableDef.regularColumns.map(_.columnName).toSet
+    val partitionKeyColumnNames = tableDef.partitionKey.map(_.columnName).toSet
+    val joinColumnNames = columns.map(_.columnName).toSet
+
+    def checkSingleColumn(column: NamedColumnRef) = {
+      if (regularColumnNames.contains(column.columnName))
+        throw new IllegalArgumentException(s"Can't pushdown join on column $column because it is not part of the PRIMARY KEY")
+      column
+    }
+
+    //Make sure we have all of the clustering indexes between the 0th position and the max requested in the join
+    val chosenClusteringColumns = tableDef.clusteringColumns
+      .filter(tableCol => columns.exists(joinCol => tableCol.columnName == joinCol.columnName))
+    if (chosenClusteringColumns.size != 0) {
+      val chosenClusteringIndexes = chosenClusteringColumns.flatMap(col => col.componentIndex)
+      val maxIndex = chosenClusteringIndexes.max
+      val requiredIndexes = 0 to maxIndex
+      val missingIndexes = requiredIndexes.toSet -- chosenClusteringIndexes.toSet
+      if (missingIndexes.size != 0) {
```

Comment:
This code seems to check if a sequence of selected columns, ordered by component index is a prefix of the clustering columns sequence. The clustering columns are already ordered by component index. If so, this can be written in a much more explicit and readable way:

```scala
if (!tableDef.clusteringColumns.startsWith(chosenClusteringColumns)) {
  val requiredColumns = tableDef.clusteringColumns.takeWhile(_.componentIndex <= chosenClusteringColumns.last.componentIndex)
  val missingColumns = requiredColumns.toSet -- chosenClusteringColumns
  // report error
}
```
You also don't need to check if any clustering column was selected at all because empty sequence is a valid prefix of every sequence.